src/shell/filesystem_shell.rs

```diff
@@ -177,23 +183,39 @@ impl Shell for FilesystemShell {
             },
             Some(v) => {
                 let target = v.as_string()?;
-                let path = PathBuf::from(self.path());
-                match dunce::canonicalize(path.join(target).as_path()) {
+                let path = PathBuf::from(self.path()).join(target.clone());
+
+                match dunce::canonicalize(path.as_path()) {
                     Ok(p) => p,
                     Err(_) => {
-                        return Err(ShellError::labeled_error(
+                        let err = Err(ShellError::labeled_error(
                             "Can not change to directory",
                             "directory not found",
                             v.span().clone(),
                         ));
+
+                        if path.ends_with("-") {
+                            match dunce::canonicalize(
+                                PathBuf::from(self.last_path.clone()).as_path(),
+                            ) {
+                                Ok(p) => p,
+                                Err(_) => {
+                                    return err;
+                                }
+                            }
+                        } else {
+                            return err;
+                        }
```

Comment:
Not very happy with the control flow here but I'm not seeing an easy way to make it clearer.